typing/types.ml

```diff
@@ -300,6 +300,12 @@ and constructor_tag =
   | Cstr_extension of Path.t * bool     (* Extension constructor
                                            true if a constant false if a block*)
 
+let equal_tag t1 t2 = match (t1, t2) with
+  | Cstr_constant i1, Cstr_constant i2 -> i2 = i1
+  | Cstr_block i1, Cstr_block i2 -> i2 = i1
+  | Cstr_extension (path1, b1), Cstr_extension (path2, b2) -> path1 = path2 && b1 = b2
+  | _ -> false
+
```

Comment:
@chambart means that there should be no final `_` case. The code should be such that, if we add a new constructor to the input datatype, we get an exhaustivity warning.

I personally structure equality tests as follows to ensure this property:

```ocaml
| K1 p, K1 q -> ...
| K1 _, _ | _, K1 _ -> false
| K2 p, K2 q -> ...
| K2 _, _ | _, K2 _ -> false
...
| Kn p, Kn p -> ...
```
(The last case does not have a catch-all sibling because all constructors have been matched.)

I'm not sure whether this generates more or less efficient matching code, so maybe it would make sense to re-run your benchmark if you change this function.